### PR TITLE
implement pick_batch to allow picking one or more batches

### PR DESCRIPTION
**Pull request by Oneplus**
This PR includes

* implementation of `PickBatch` function and  `pick_batch` expression.
* test case for pick_batch
* an example that doing document classification on IMDB dataset.

Example:
For a Tensor with 4 batches:
```
[[0 1 2] [3 4 5]]
[[6 7 8] [9 10 11]]
[[12 13 14] [15 16 17]]
[[18 19 20] [21 22 23]]
```
`pick_batch(t, 1)` will return a Tensor of
```
[[6 7 8] [9 10 11]]
```
`pick_batch(t, {1, 2})` will return a Tensor of
```
[[6 7 8] [9 10 11]]
[[12 13 14] [15 16 17]]
```

Details:
Implementation basically follows the implementation of `PickElement` function which broadcast the Tensor with `tb()` and select the last row (batch row) with `chip` function.

### CODE REVIEW

File: dynet/expr.h [hunk 1]
```diff
@@ -1354,6 +1354,33 @@ Expression pickrange(const Expression& x, unsigned v, unsigned u);
...
+ * \param v The beginning index
+ *
+ * \return The expression of picked batch. The picked batch is a tensor
+ *         whose `bd` equals to one.
+ */
+Expression pick_batch(const Expression& x, unsigned v);
+
+/**
+ * \ingroup flowoperations
+ * \brief Batched pick
```
Comment (neubig) [inline dynet/expr.h]: I think this doc is wrong, except for the return statement.

File: dynet/expr.cc [hunk 2]
```diff
@@ -115,6 +115,9 @@ Expression pick(const Expression& x, const vector<unsigned> & v, unsigned d) { r
 Expression pick(const Expression& x, const unsigned* pv, unsigned d) { return Expression(x.pg, x.pg->add_function<PickElement>({x.i}, pv, d)); }
 Expression pick(const Expression& x, const vector<unsigned> * pv, unsigned d) { return Expression(x.pg, x.pg->add_function<PickElement>({x.i}, pv, d)); }
 
+Expression pick_batch(const Expression& x, unsigned v) { return Expression(x.pg, x.pg->add_function<PickBatch>({x.i}, v)); }
+Expression pick_batch(const Expression & x, const std::vector<unsigned>& v) { return Expression(x.pg, x.pg->add_function<PickBatch>({x.i}, v)); }
```
Comment (neubig) [inline dynet/expr.cc]: Let's name this pick_batches to be clear that it's plural.

File: dynet/nodes.cc [hunk 3]
```diff
@@ -1549,6 +1549,42 @@ void PickRange::backward_dev_impl(const MyDevice & dev,
...
+  } else {
+    DYNET_ASSERT(pvals != nullptr, "Neither single nor vector of elements available in PickBatch::forward");
+    if (pvals->size() != fx.d.batch_elems())
+      DYNET_INVALID_ARG("In PickBatch::forward, number of elements in the passed-in index vector (" << pvals->size() <<
+                        ") did not match number of elements in mini-batch elements in expression (of dimension" << fx.d << ")");
+    for (unsigned b = 0; b < pvals->size(); ++b) {
+      if ((*pvals)[b] >= xs[0]->d.nd)
+        DYNET_INVALID_ARG("PickBatch::forward_impl requested element " << (*pvals)[b]
+                          << " from a batch size of " << xs[0]->d.nd);
+      fx.tb<2>().chip<2>(b).device(*dev.edevice) = xs[0]->tb<3>().chip<3>((*pvals)[b]);
```
Comment (neubig) [inline dynet/nodes.cc]: Similarly `tbvec()` and `tbvec()`.

File: tests/test-nodes.cc [hunk 4]
```diff
@@ -902,6 +902,15 @@ BOOST_AUTO_TEST_CASE( pick_batch_gradient ) {
   BOOST_CHECK(check_grad(mod, z, 0));
 }
 
+// Expression pick_batch(const Expression& x, unsigned v);
+BOOST_AUTO_TEST_CASE( pick_batch_gradient1 ) {
```
Comment (neubig) [inline tests/test-nodes.cc]: Can we get a test for `pick_batches()` as well?

File: dynet/nodes.cc (hunk 3 above)
Comment (neubig) [inline dynet/nodes.cc]: Woops, sorry, my first comment disappeared. In all places, `tb<3>()` etc. can probably be replaced with `tbvec()` or `tvec()`

File: tests/test-nodes.cc (hunk 4 above)
Comment (Oneplus) [inline tests/test-nodes.cc]: solved in the new commit

**Pull request merged**